### ISSUE TITLE
WebGL error texture bound to texture unit 1 is not renderable

### ISSUE THREAD

**Issue by viraxslot** [molstar/molstar#319]
Hi, guys. I faced this error in our React App:
```
[.WebGL-0x7f872a918c00]RENDER WARNING: texture bound to texture unit 1 is not renderable. It might be non-power-of-2 or have incompatible texture filtering (maybe)?
```
I'm using PluginContext like in examples in the repo. 
[Here on stackoverflow](https://stackoverflow.com/questions/19722247/webgl-wait-for-texture-to-load/19748905#19748905) proposed to "make a 1x1 texture at creation time" (similar error but not for molstar). 
I wanna try this workaround.

So my question is: how can I do that? Should I use `applyPreset` function or something else?
Maybe you have a different solution for my problem. 

Thanks!

**Comment by dsehnal** [comment]
Hi,
what input does that happen on? I can't recall ever seeing this error. Could you help us identify what causes this so that we can solve this internally?

**Comment by viraxslot** [comment]
@dsehnal Hi, David. I can provide input a little bit later.

System information:
```
OS: macOS Catalina, 10.15.7 (19H1519)
Browser: Chrome Version 96.0.4664.110 (Official Build) (x86_64)

Model Name: MacBook Pro
Model Identifier: MacBookPro16,1
Processor Name: 6-Core Intel Core i7
Processor Speed: 2,6 GHz
```

I tried to reproduce this issue on Firefox - works just fine.
So maybe it's not about the data.

**Comment by dsehnal** [comment]
Yeah, that's quite an outdated OS and I know there have been some issues with Chrome 96. Browsers sometimes break things in between versions which get later fixed on their own.

**Comment by arose** [comment]
Hi, I see that message on browserscope for your versions. However things seem to render fine. After all they are warnings not errors and should not break your app.

![image](https://user-images.githubusercontent.com/272250/146992408-5e7d3732-9dc6-4e68-845a-0e008d131351.png)

**Comment by viraxslot** [comment]
@arose Hi, thanks for reproducing my problem. You're right there are no problems with rendering, but is there any workaround to get rid of these messages? They prevent, for example, during the test run, because it's really hard to find useful information in the console.

**Comment by azangru** [comment]
@arose  @dsehnal  I may have found some additional clues for debugging. It is sufficient to just start a canvas3D to get the warnings about non-renderable textures:

```
[.WebGL-0x1400d29c00]RENDER WARNING: texture bound to texture unit 2 is not renderable. It might be non-power-of-2 or have incompatible texture filtering (maybe)?
```

It seems that when the canvas is initialised, some of the textures for which the [`.bind` method](https://github.com/molstar/molstar/blob/master/src/mol-gl/webgl/texture.ts#L340-L343) is executed have a width and a height of 0. There is something not quite right about these textures, and the WebGL context in Chrome on MacOS complains that they are not renderable. An early return from the `bind` and the `unbind` function if the width and the height of a texture are 0 gets rid of these warning.

It still doesn't help against the last warning on the screenshot above, `RENDER WARNING: there is no texture bound to the unit 0`; but this one appears only after a molecule representation is added to the canvas.

**Comment by azangru** [comment]
Last piece of the puzzle: render warnings complaining that `there is no texture bound to the unit 0;` are indeed caused by [passing `null` to `gl.bindTexture` in the bind method of the null texture](https://github.com/molstar/molstar/blob/master/src/mol-gl/webgl/texture.ts#L432). When I replaced that line with the code that generates a 1x1px texture, as suggested in the [SO discussion](https://stackoverflow.com/questions/19722247/webgl-wait-for-texture-to-load/19748905#19748905) mentioned above:

```
var tex = gl.createTexture();
gl.bindTexture(gl.TEXTURE_2D, tex);
gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, 1, 1, 0, gl.RGBA, gl.UNSIGNED_BYTE,
              new Uint8Array([255, 0, 0, 255])); // red
```

the render warning went away.

So, to summarise:
- escaping early from the `.bind` method for textures with a zero width and height takes care of the first warning (`RENDER WARNING: texture bound to texture unit 2 is not renderable. It might be non-power-of-2 or have incompatible texture filtering (maybe)?`)
- starting with a 1x1px texture in the `.bind` method of the null texture gets rid of the second warning (`RENDER WARNING: there is no texture bound to the unit 0`)

**Comment by arose** [comment]
Thank you very much for the investigation @azangru. I will see how to best apply your findings.

**Comment by arose** [comment]
@azangru @viraxslot I pushed some changes that work for me on macOS Catalina/Chrome 96 on browserstack. Please have a look and let me know if you see anymore warnings.

The "texture not renderable" error originated from the SMAA code where we load texture data async so they are uninitialized and of size zero for a short while (https://github.com/molstar/molstar/blob/master/src/mol-canvas3d/passes/smaa.ts#L193-L209 )

**Comment by azangru** [comment]
@arose I can confirm that both warnings are now gone. Tested locally on MacOS Mojave (👴) on Chrome 98.

However, while testing, I noticed something interesting. I have a playground where I am testing different Molstar plugin instances — with the default UI, without the default UI, and canvas3D; and while switching between the instances to check for the warnings, I observed the following:

1. Switching from plugin without default UI to canvas3d and then back to plugin without default UI prints the following warning in the console:

```
Symbol 'computed.accessible-surface-area.is-buried' already added. Call removeSymbol/removeCustomProps re-adding the symbol.
```

2. Same scenario (plugin without UI -> canvas 3d -> plugin without UI) followed by window resizing prints the following error:

```
Uncaught (in promise) Error: empty textures are not allowed
    at Object.define (texture.ts:264:19)
    at PostprocessingPass2.setSize (postprocessing.ts:362:35)
    at DrawPass2.setSize (draw.ts:161:33)
    at Passes2.updateSize (passes.ts:26:19)
    at handleResize (canvas3d.ts:679:20)
    at render (canvas3d.ts:380:17)
    at draw (canvas3d.ts:426:17)
    at Object.tick (canvas3d.ts:446:13)
    at PluginAnimationLoop2.<anonymous> (animation-loop.ts:21:31)
    at step (tslib.es6.js:102:23)
```

3. Resizing the viewport when the plugin without the built-in UI is enabled always results in the growing of the canvas

https://user-images.githubusercontent.com/6834224/154852039-573f4046-55dd-4695-9491-60a48336e5e9.mp4

I understand that all three of the above observations have nothing to do with the original issue. First two observations are probably because the plugin instance isn't properly uninitialized when I switch to a different view. Do you think any of the above deserve a dedicated issue, or is it just me not using the API properly (I do run `plugin.clear()` before unmounting the component, but it is not sufficient to unregister the plugin)?

**Comment by arose** [comment]
@azangru thanks for checking. Would you mind opening new issues for the other points?

**Comment by azangru** [comment]
I wish I read the docs more closely 🙂  I've opened an [issue](https://github.com/molstar/molstar/issues/379), and then found the answer in the docs. Turns out, the cleanup method I was after is `plugin.dispose`. Now that I am calling it on unmount, the console is clear.
